### ISSUE TITLE
Wildcards on IP Literals

### ISSUE THREAD

**Issue by ericlaw1979** [vcsjones/FiddlerCertGen#12]
A quick read of https://github.com/vcsjones/FiddlerCertGen/commit/28e1acd9dfe18c25895415800ff9ba52402de7a0 suggests that your code will attempt to generate a wildcard like `*.123.123.123` for a host that is an IPv4 literal `123.123.123.123`.

**Comment by vcsjones** [comment]
This is fixed, indirectly. Thanks for pointing it out. The cert generator will now attempt to parse the host name as an IP Address. If it is (IPv4 or IPv6) it will use an "IP Address" in the subject alternative name instead of a DNS Name. The wildcard logic is only applicable to DNS Names, so we won't generate a wild card cert for IP addresses anymore.

**Comment by ericlaw1979** [comment]
I know the spec calls for it, but do all clients support SANs of type IPAddress, or do they look for DNSName even for ip literals?

**Comment by vcsjones** [comment]
> do all clients support SANs of type IPAddress

No. It appears the newer ones (Chrome 46, IE 11, etc) do.

Some browsers accept the IP Address in the CN, some don't. For DNS Name, I am not sure about IP literals in DNS Name. Newer browsers do _not_ accept it, but some older ones might, which I have not exhaustively tested.

**Comment by ericlaw1979** [comment]
Newer browsers _don't_ accept ipliterals as DNS names? This is going to get very interesting indeed.

**Comment by vcsjones** [comment]
No. Chrome does not like that. It accepts subjAltName with IP Address, but not DNS Name.

Internet Explorer 11 on Windows 10 accepts it.
<img width="1095" alt="screen shot 2015-10-30 at 6 33 22 pm" src="https://cloud.githubusercontent.com/assets/361677/10859615/c5fcef02-7f35-11e5-8ff0-262eff84b135.png">
<img width="705" alt="screen shot 2015-10-30 at 6 36 44 pm" src="https://cloud.githubusercontent.com/assets/361677/10859616/c615b848-7f35-11e5-906f-d57138370cd1.png">

**Comment by vcsjones** [comment]
Seems like it's probably just a good idea to do both, support DNS Name and IP Address in the hopes of covering as many browsers as we can by IP address literals.

**Comment by ericlaw1979** [comment]
It turns out that IE/CAPI doesn't do ipv6Literal matching correctly in subjectAltNames. I've changed the certenroll provider to detect if the target is `AddressFamily.InterNetworkV6` and if so, it adds two SubjectAltNames, one of type `IPAddress` with the correct value and a hackish extra of type `DNSName` with value `"[" + ipv6address + "]"`;

**Comment by vcsjones** [comment]
Interesting. I'll make a fix.

**Comment by vcsjones** [comment]
I made a fix for this in ee717347c1b44f8a42f52c4012838f8ef2e95b3a.

I did not use `$"[{address.ToString()}]"`, rather I used the value that is given in the Host (`sHostname`). My fear with using `ToString()` on the address is that it will expand the IP address from any short forms. For example, if the address `::1` is what gets typed in the browser, `ToString` will turn that into `0000:0000:0000:0000:0000:0000:0000:0001`, and that won't be the same as `::1` if a string comparison is really being done by CAPI.
